feat: Add a display name parameter

Glutexo opened the pull request:
Enhancement:
A new _rhc_insights.display_name_ parameter.

Reason:
The display name is a value managed by Insights Client. It is visible in the Host-Based Inventory and is used as the main visible name in the UI to identify a host and search for it.

Result:
The new _rhc_insights.display_name_ parameter sets the display name of the host. This value is written to the Insights Client config file and also set directly by a `--display-name` command. When marked as absent or on unregistration, the value is removed from the config file.

Issue Tracker Tickets (Jira or BZ if any):
https://issues.redhat.com/browse/CCT-13

richm commented on tasks/insights-client.yml:
```diff
@@ -94,6 +94,62 @@
+        regexp: "^display_name="
+        insertafter: "#display_name="
+        line: display_name={{ rhc_insights.display_name }}
+    - name: Update display namein inventory
```
```suggestion
    - name: Update display name in inventory
```

Glutexo replied:
Also probably Inventory should be written with a capital I. But keeping that lowercase for consistency with the Ansible Host tasks descriptions.

ptoscano commented on README.md:
````diff
@@ -173,6 +173,20 @@ Possible values of this variable:
+      display_name: "Example Host"
+```
+
+Configures the display name with a custom value for the system record in Host Based Inventory (HBI). This display name is only used to identify the host in the Invenotry.
````
typo "Invenotry"

Glutexo replied:
If the only comment on the description is a typo, it’s good news! 🥳

ptoscano reviewed (commented):
seems a good start; please don't forget a test for it

Glutexo commented:
Fixed the typos. Going to add the tests.

Glutexo commented:
I added the tests, so I am opening the pull request for a review.

DuckBoss commented:
I had a quick look over -- make sure the tests are referencing `display_name` instead of `ansible_host`. I know the tests are near-identical to `ansible_host` tests but I'm assuming leaving the fields as `ansible_host` was unintentional.

I can test these changes shortly after you push the changes :+1:

DuckBoss reviewed (changes requested):
I had a quick look over -- make sure the tests are referencing `display_name` instead of `ansible_host`. I know the tests are near-identical to `ansible_host` tests but I'm assuming leaving the fields as `ansible_host` was unintentional.

I can test these changes shortly after you push the changes :+1:

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+    - name: Setup Insights
+      import_tasks: tasks/setup_insights.yml
+
+    - name: Test ansible_host
```
```suggestion
    - name: Test display_name
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+
+    - name: Test ansible_host
+      block:
+        - name: Add ansible_host and register insights
```
```suggestion
        - name: Add display_name and register insights
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+                username: "{{ lsr_rhc_test_data.reg_username }}"
+                password: "{{ lsr_rhc_test_data.reg_password }}"
+            rhc_insights:
+              ansible_host: host
```
```suggestion
              display_name: friendly_hostname
```

Glutexo replied:
friendly_hostname 😊

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+              insecure: "{{ lsr_rhc_test_data.candlepin_insecure }}"
+            rhc_baseurl: "{{ lsr_rhc_test_data.baseurl | d(omit) }}"
+
+        - name: Check ansible_host is set to 'host' in config file
```
```suggestion
        - name: Check display_name is set to 'friendly_hostname' in config file
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+
+        - name: Check ansible_host is set to 'host' in config file
+          command:
+            grep -ixq "^ansible_host=host" {{ __rhc_insights_conf }}
```
```suggestion
            grep -ixq "^display_name=friendly_hostname" {{ __rhc_insights_conf }}
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+            grep -ixq "^ansible_host=host" {{ __rhc_insights_conf }}
+          changed_when: true
+
+        - name: Test ansible_host changed value after registration
```
```suggestion
        - name: Test display_name changed value after registration
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+
+        - name: Test ansible_host changed value after registration
+          block:
+            - name: Change ansible host to 'new-host'
```
```suggestion
            - name: Change display_name to 'new-host'
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+                name: linux-system-roles.rhc
+              vars:
+                rhc_insights:
+                  ansible_host: new-host
```
```suggestion
                  display_name: new-host
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+                rhc_insights:
+                  ansible_host: new-host
+                  remediation: absent
+            - name: Check ansible_host is set to 'new-host' in config file
```
```suggestion
            - name: Check display_name is set to 'new-host' in config file
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+                  remediation: absent
+            - name: Check ansible_host is set to 'new-host' in config file
+              command:
+                grep -ixq "^ansible_host=new-host" {{ __rhc_insights_conf }}
```
```suggestion
                grep -ixq "^display_name=new-host" {{ __rhc_insights_conf }}
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+              command:
+                grep -ixq "^ansible_host=new-host" {{ __rhc_insights_conf }}
+              changed_when: false
+            - name: Change ansible host to a null value (noop)
```
```suggestion
            - name: Change display_name to a null value (noop)
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+                name: linux-system-roles.rhc
+              vars:
+                rhc_insights:
+                  ansible_host: null
```
```suggestion
                  display_name: null
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+                rhc_insights:
+                  ansible_host: null
+                  remediation: absent
+            - name: Check ansible_host has not changed in config file
```
```suggestion
            - name: Check display_name has not changed in config file
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+                  remediation: absent
+            - name: Check ansible_host has not changed in config file
+              command:
+                grep -ixq "^ansible_host=new-host" {{ __rhc_insights_conf }}
```
```suggestion
                grep -ixq "^display_name=new-host" {{ __rhc_insights_conf }}
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+                grep -ixq "^ansible_host=new-host" {{ __rhc_insights_conf }}
+              changed_when: false
+
+        - name: Test ansible_host set to an absent value
```
```suggestion
        - name: Test display_name set to an absent value
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+
+        - name: Test ansible_host set to an absent value
+          block:
+            - name: Set ansible host to an absent value
```
```suggestion
            - name: Set display_name to an absent value
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+                name: linux-system-roles.rhc
+              vars:
+                rhc_insights:
+                  ansible_host: {state: absent}
```
```suggestion
                  display_name: {state: absent}
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+                rhc_insights:
+                  ansible_host: {state: absent}
+                  remediation: absent
+            - name: Check ansible_host is not present in config file
```
```suggestion
            - name: Check display_name is not present in config file
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+            - name: Check ansible_host is not present in config file
+              lineinfile:
+                path: "{{ __rhc_insights_conf }}"
+                regexp: "^ansible_host="
```
```suggestion
                regexp: "^display_name="
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+                regexp: "^ansible_host="
+                state: absent
+              check_mode: true
+              register: __test_ansible_host_absent
```
```suggestion
              register: __test_display_name_absent
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+                state: absent
+              check_mode: true
+              register: __test_ansible_host_absent
+              failed_when: __test_ansible_host_absent.found
```
```suggestion
              failed_when: __test_display_name_absent.found
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+              register: __test_ansible_host_absent
+              failed_when: __test_ansible_host_absent.found
+
+        - name: Test ansible_host is removed during insights unregistration
```
```suggestion
        - name: Test display_name is removed during insights unregistration
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+
+        - name: Test ansible_host is removed during insights unregistration
+          block:
+            - name: Set ansible_host in config file
```
```suggestion
            - name: Set display_name in config file
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+                name: linux-system-roles.rhc
+              vars:
+                rhc_insights:
+                  ansible_host: unreg-test
```
```suggestion
                  display_name: unreg-test
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+                rhc_insights:
+                  ansible_host: unreg-test
+                  remediation: absent
+            - name: Check ansible_host is set in config file
```
```suggestion
            - name: Check display_name is set in config file
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+                  remediation: absent
+            - name: Check ansible_host is set in config file
+              command:
+                grep -ixq "^ansible_host=unreg-test" {{ __rhc_insights_conf }}
```
```suggestion
                grep -ixq "^display_name=unreg-test" {{ __rhc_insights_conf }}
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+              vars:
+                rhc_insights:
+                  state: absent
+            - name: Check ansible_host is not present in config file
```
```suggestion
            - name: Check display_name is not present in config file
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+            - name: Check ansible_host is not present in config file
+              lineinfile:
+                path: "{{ __rhc_insights_conf }}"
+                regexp: "^ansible_host="
```
```suggestion
                regexp: "^display_name="
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+                regexp: "^ansible_host="
+                state: absent
+              check_mode: true
+              register: __test_ansible_host_removed
```
```suggestion
              register: __test_display_name_removed
```

richm commented on tests/tests_insights_display_name.yml:
```diff
@@ -0,0 +1,116 @@
+                state: absent
+              check_mode: true
+              register: __test_ansible_host_removed
+              failed_when: __test_ansible_host_removed.found
```
```suggestion
              failed_when: __test_display_name_removed.found
```

Glutexo commented:
🤦🏻‍♂️ I did those changes already, but of couse I didn’t add them to Git before committing. Thank you for spotting that! Pushed the fixed test file.

Glutexo commented:
Filled in the pull request description.

richm commented:
[citest]

DuckBoss commented on tasks/insights-client.yml:
```diff
@@ -94,6 +94,62 @@
+        regexp: "^display_name="
+        state: absent
+    - name: Remove display name in inventory
+      shell: insights-client --display-name= & wait
```
Resetting the display name using this command doesn't seem to be working as expected in insights-client. It just starts another insights collection instead of reverting the display-name back to the system name which causes this to silently fail.

I'll have a look at the insights-client code that handles this, because it looks like it doesn't handle empty values being passed through like the `ansible-host` command.

DuckBoss replied:
Oops didn't notice Pino already posted about this issue!

richm commented:
> The implementation seems mostly OK, and the test passes when running it against Insights.
> 
> The problem I see is when unsetting the display name (i.e. `{state: absent}`): whereas the configuration is properly updated, the display name in Inventory is not changed. In particular, it looks like `insights-client --display-name=` does not do much, as it is considered `None` by insights-core, and thus it is a no-op.
> 
> I see few possibilities for this:
> 
>     1. fix insights-core to allow an empty string for the display name as synonym for "use the system hostname"
> 
>     2. manually pass an hostname here in the role by getting it manually; the logic in insights-core is IMHO not easy to duplicate manually [1], so it would be either
>        
>        * unconditionally use something like `hostname -f`
>        * put the logic below in a small private module for this role
> 
>     3. "disallow" unsetting the display name: an empty string would behave as if `display_name:` was not passed (i.e. not changing it), and document this situation
> 
> 
> [1] https://github.com/RedHatInsights/insights-core/blob/396355192c229982f5cb55288b92c21a82151240/insights/util/hostname.py
> 
> Personally, I'd go for (2a) or (3).
> 
> @Glutexo @richm What do you think?

2. seems the best option to me

ptoscano commented:
>     2. seems the best option to me

(2) as in "the simple `hostname` way", or the small helper private module? (I put each as possible option, not both together)

richm commented:
> > ```
> > 2. seems the best option to me
> > ```
> 
> (2) as in "the simple `hostname` way", or the small helper private module? (I put each as possible option, not both together)

As in the simple `hostname` way

richm commented:
> @ptoscano I prefer option (3) and updating insights-core afterwards to have similar functionality to `--ansible-host` when given empty values. Personally, id prefer not to add custom logic in the role to handle this with option (2) in this case.

3 is also fine with me

ptoscano commented:
> @ptoscano I prefer option (3) and updating insights-core afterwards to have similar functionality to `--ansible-host` when given empty values. Personally, id prefer not to add custom logic in the role to handle this with option (2) in this case.

Makes sense to me; let's go for this then.

Glutexo commented:
Oh, I didn’t know `insights-client --display-name=` does nothing. That’s exatly the think that I would have caught during proper manual testing. Nice catch! ⚾️

I completely agree with @DuckBoss here on going with way ⑶, treating an empty string as a _null_ value; and of course documenting it. I believe that the logic of determining a hostname (way ⑵) doesn’t belong to the role: it’s reponsibility of the Client, which doesn’t support this (yet).

It makes sense to me to add the feature to the Client, so the behavior is equivalent to that of `--ansible_host`, which would be reasonable even without the RHC role. Once it’s implemented, we’d revisit the role and make it use the new feature. So, way ⑴, but as a separate step.

We have a consensus of alll four of us on the way to go. Going to do that.

Glutexo commented:
I removed the subtask trying to unset the display name by running `insights-client display_name=` and updated the README. The changes are pushed to this PR.

richm commented:
[citest]

richm commented:
lgtm - @ptoscano @DuckBoss ?

DuckBoss reviewed (approved):
LGTM, let's revisit wrt `--display-name` after changes to insights-client.

ptoscano reviewed (changes requested):
> * `{state: absent}`: the display name is unset in the insights-client config file; please note that Host Based Inventory is not updated

I don't think this is a good approach: the user will see that the insights-client configuration was changed (good), but not what is in Inventory (bad), leading the users to think that their operation succeeded. If it is not possible to "unset" (i.e. reset to the hostname) the display name, then the role ought to not allow the users to do this.

Hence:
- drop the handling of `{state: absent}` altogether
- handle empty values as `null`, documenting this behaviour as limitation in the underlying tooling (`insights-client`)

ptoscano reviewed (changes requested):
@Glutexo not sure why I'm requesting to review again; see my previous comment on 'state: absent'

Glutexo commented:
I spotted an issue when `display_name` is set to an empty string:

- `display_name=` gets written into the config file, causing the display name not to be updated. It is equivalent to a missing value in the file.
- The Client is run with `--display_name=`, which is ignored and causes a regular connection to run. That is an unwanted and resource-heavy process.

It is the same problem as I found in https://github.com/linux-system-roles/rhc/pull/155#pullrequestreview-1882483685. Since this pull request already has an approval from Jason, it may make sense to merge it as it is and fix the bug for both parameters at the same time in a follow-up pull request. If you think it’s worth including in this one, I‘ll push the fix.

Glutexo commented:
@ptoscano

The `{state: absent}` is however the only way how to remove the value from the config file, causing it at least not to be overwritten on collection. It is not an ideal solution, but as long as it’s documented it at least provides the user with an additional option.

Moreover, the non-ideal state that may cause confusion to the uses (although the real behavior is documented) would only last until the Client is patched to support display name removal. That is true for the absence of `{state: absent}` too, though.

Still, I’d be more inclined to offer a way to clean the configuration, when there is one to add it.

What do you think, Pino?

ptoscano commented:
> The `{state: absent}` is however the only way how to remove the value from the config file, causing it at least not to be overwritten on collection.
> [...]
> Still, I’d be more inclined to offer a way to clean the configuration, when there is one to add it.

I'm not sure what would be the point of removing the value from the config file, if that will not reflect what is the state in Inventory.

Please see my previous comment which explains the situation as I see it: https://github.com/linux-system-roles/rhc/pull/166#pullrequestreview-1883619178

> Moreover, the non-ideal state that may cause confusion to the user (although the real behavior is documented) would only last until the Client is patched to support display name removal. That is true for the absence (😃) of `{state: absent}` too, though.

Yes, and I do not think the role ought to do operations that do not actually work. When `insights-client` (or actually `insights-core`) will be changed to allow resetting/unsetting the display name, then the role may be adapted to make use of that.

Glutexo commented:
Since Pino requested changes on the pull request, I believe I can fix the empty string issue in this pull request. Pushed a new commit that causes the empty string to be treated as an empty value: the display name is ignored in that case.

Glutexo commented:
Pushed a patch removing the `{state: absent}` value. If a user still tries to provide such value, it is converted to a string by Python (to `{'state': 'absent'}`, which causes error. I believe this is not something we need to cover, as this happens to other string values too.

richm commented on README.md:
```diff
@@ -173,6 +173,19 @@ Possible values of this variable:
+Possible values of this variable:
+
+* `null` or an empty string: the display name is not changed
+* any other string value: the display name is changed in Host Based Inventory (HBI).
```
Do we need a `NOTE: Once set, the display name cannot be removed or reverted back to the original value.  You can only set a new value.`
?

ptoscano replied:
Strictly speaking, there is always a display name set: in particular, the default one that `insights-client` sets (if not specified otherwise) is the hostname.

I would either write something slightly more verbose along the lines of

> Note: `insights-client` sets the default display name on registration as the hostname of the system, in case it is not explicitly configured. When changing the display name manually using this variable, it is not possible to "revert it" automatically to be the hostname, and thus it must be done manually if wanted.

... or nothing at all.

Glutexo replied:
Is is true that it’s the `insights-client` that sets the display name to the hostname, or is it sent empty and set automatically in the cloud?

Glutexo replied:
Answer: It’s not the Client that sets the default display name. If not set, it doesn’t send it at all:

- no _display_name_ field in the canonical facts JSON;
- no _display_name_ file in the archive.

That means that the system gets its display name when processed in the Platform, not on the machine.

ptoscano commented:
> The role is not primarily intended to edit the Client config file. It uses the Client as a tool to perform changes in the Inventory and changes the config file so the configuration is consistent with the new state and doesn’t get reverted on the next collection.

Yes and no.

Take a step back: a role/playbook/module/etc is nothing more than a glorified/structured way to do operations and check their results. Whatever is done with Ansible, it can be run manually to get the same result.

In case of this role, its purpose is to create and maintain the connectivity of systems to Red Hat or Satellite, Insights included. If to do so some configuration files must be edited, then they are edited; if some command must be run, they are run; and so on.

In the specific case of this option, we need to ensure that `insights-client` will keep reporting the configured display name to the connected Inventory. Since querying manually Inventory would sort of "go around" the purpose of `insights-client`, then we tell `insights-client` which display name to use by writing it in the config file, and then running `insights-client` as needed to ensure Inventory is up-to-date with the wanted configuration.

ptoscano reviewed (changes requested):
The changes seem fine to me now, thanks!

The only thing to check would be what Rich mentioned (and I added some notes) on the documentation about the lack of "unset"; let's include what I wrote (or something along those lines), if that looks OF for you.

Since the time for merging is closer, please rebase and squash the commits into one with a proper commit message for it.

Glutexo commented:
I added a note on the default display name value to the documentation. I found the suggested wording a bit too verbose, so I tried to shorten it a bit.

ptoscano commented:
> please rebase and squash the commits into one with a proper commit message for it.

this still needs to be done

Glutexo commented:
Squashed the branch into a single commit and force-pushed.

richm commented:
[citest]

Glutexo commented:
I realized that I added the empty string condition to fix a buggy behavior, but without a test.

I added a test verifying that setting _rhc_insights.display_name_ to an empty string doesn’t change the value in the config file, just as a _null_ doesn’t. Without the added condition, the test would fail, because the empty string would change the line in the config file to `display_name=`.

I didn’t test calling the `insights-client`, because that is not tested at all by any of the existing tests.

ptoscano commented:
[citest]

Glutexo commented:
From what I found that the Insights Client doesn’t send an empty display name, I am a bit afraid that adding support for the Client to reset the value in the Inventory would require a change on the Platform side. I will however verify, what is the actual behavior of sending various empty values of the display name.

ptoscano reviewed (approved):
LGTM now -- thanks for the work on this!
